[Security] Add missing upgrade note about GuardAuthenticationListener-related deprecation

DonCallisto opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.0
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   |
| Fixed tickets |
| License       | MIT
| Doc PR        | 

This https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php#L127 was not reported.

nicolas-grekas commented:
Should we tell about any corresponding deprecation in UPGRADE-3.4.md also?

DonCallisto commented:
I'm not sure what you mean but to me it's pretty fine to enhance the upgrade document from this major version.

nicolas-grekas commented:
This should be merged on branch 3.4, where we advertise for changes needed to prepare for 4.0.
I'm surprised there is no corresponding changes on UPGRADE-3.4.md. Could anyone have a look?

nicolas-grekas commented:
So, following the link in the description gives the answer: the depreciation should be advertised in UPGRADE-3.4.md for sure.

DonCallisto commented:
I'm still not sure as this is deprecated since 3.4 but I should not care about this *unless* upgrading to 4.
However if that's the practice followed here, I'll be happy to change the base of this PR, _just let me know_.
After this I'll change what @iltar reviewed in order to make the docs consistent.

nicolas-grekas commented:
I do confirm this should be on 3.4 :)

DonCallisto committed:
Update UPGRADE-4.0.md

This https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php#L127 was not reported.

DonCallisto commented:
@nicolas-grekas I've changed the PR base branch
@iltar I've fixed what you've requested

nicolas-grekas commented:
Thank you @DonCallisto 
Can you please also update UPGRADE-3.4.md?
I invite you to compare both files (4.0 and 3.4) to figure out what I'm talking about since my first comment :)

nicolas-grekas commented:
And src/Symfony/Component/Security/CHANGELOG.md also should be updated BTW!

DonCallisto commented:
Ok, thank you for this advices. I'll take a look at them ASAP.

DonCallisto commented:
@nicolas-grekas updated. Please check if fits well and let me know if anything else has to be changed.

nicolas-grekas commented on CHANGELOG-3.4.md:
```diff
@@ -726,4 +726,6 @@ To get the diff between two versions, go to https://github.com/symfony/symfony/c
  * feature #22828 [Finder] Deprecate FilterIterator (ogizanagi)
  * feature #22826 [Validator] improve strict option value deprecation (xabbuh)
+ * feature #16835 [Security] Deprecated `GuardAuthenticatorInterface` in favor of `AuthenticatorInterface` (Amo)
+ * feature #16835 [Security] When extending `AbstractGuardAuthenticator` it's no longer allowed to return `null` from `getCredentials()` (Amo)
```
I really meant `src/Symfony/Component/Security/CHANGELOG.md`
this one is autogenerated so this should be reverted :)

nicolas-grekas commented on UPGRADE-3.4.md:
```diff
@@ -344,6 +344,9 @@ Security
  * The `GuardAuthenticatorInterface` has been deprecated and will be removed in 4.0.
    Use `AuthenticatorInterface` instead.
+   
+ * When extending `AbstractGuardAuthenticator` it's no longer allowed to return `null` from `getCredentials()`.
```
~it's no longer allowed~
it's deprecated

DonCallisto commented:
Now it's hopefully fine

nicolas-grekas reviewed (approved):
all good on my side :)

fabpot commented on UPGRADE-4.0.md:
```diff
@@ -758,6 +758,9 @@ Security
  * The `GuardAuthenticatorInterface` interface has been removed.
    Use `AuthenticatorInterface` instead.
+   
+ * When extending `AbstractGuardAuthenticator` it's deprecated to return `null` from `getCredentials()`.
```
In 4.0, it's not deprecated anymore. So, it should probably be something like "When extending `AbstractGuardAuthenticator`,  `getCredentials()` cannot return `null` anymore, return `false` from `supports()` if no credentials available instead."

fabpot commented on src/Symfony/Component/Security/CHANGELOG.md:
```diff
@@ -16,6 +16,8 @@ CHANGELOG
  * deprecated HTTP digest authentication
  * Added a new password encoder for the Argon2i hashing algorithm
  * deprecated `GuardAuthenticatorInterface` in favor of `AuthenticatorInterface`
+ * deprecated to return `null` from `getCredentials()` in classes that extend `AbstractGuardAuthenticator`. Return `false` 
```
CS: should be wrapper at around 80 chars.

DonCallisto commented:
@fabpot fixed

fabpot commented:
Thank you @DonCallisto.
